// apollo-server-plugin-operation-registry/dist/common.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.signatureForLogging = exports.getOperationManifestUrl = exports.getStorageSecretUrl = exports.getStoreKey = exports.urlStorageSecretBase = exports.urlOperationManifestBase = exports.fakeTestBaseUrl = exports.envOverrideStorageSecretBaseUrl = exports.envOverrideOperationManifest = exports.pluginName = void 0;
exports.pluginName = require('../package.json').name;
exports.envOverrideOperationManifest = 'APOLLO_OPERATION_MANIFEST_BASE_URL';
exports.envOverrideStorageSecretBaseUrl = 'APOLLO_STORAGE_SECRET_BASE_URL';
exports.fakeTestBaseUrl = 'https://fake-host-for-apollo-op-reg-tests';
exports.urlOperationManifestBase = ((_a = process.env[exports.envOverrideOperationManifest]) === null || _a === void 0 ? void 0 : _a.replace(/\/$/, '')) ||
    process.env['__APOLLO_OPERATION_REGISTRY_TESTS__'] === 'true'
    ? exports.fakeTestBaseUrl
    : 'https://operations.api.apollographql.com';
exports.urlStorageSecretBase = ((_b = process.env[exports.envOverrideStorageSecretBaseUrl]) === null || _b === void 0 ? void 0 : _b.replace(/\/$/, '')) ||
    process.env['__APOLLO_OPERATION_REGISTRY_TESTS__'] === 'true'
    ? exports.fakeTestBaseUrl
    : 'https://storage-secrets.api.apollographql.com';
const getStoreKey = (signature) => `${signature}`;
exports.getStoreKey = getStoreKey;
function getStorageSecretUrl(graphId, apiKeyHash) {
    return `${exports.urlStorageSecretBase}/${graphId}/storage-secret/${apiKeyHash}.json`;
}
exports.getStorageSecretUrl = getStorageSecretUrl;
function getOperationManifestUrl(graphId, storageSecret, graphVariant) {
    return `${exports.urlOperationManifestBase}/${graphId}/${storageSecret}/${graphVariant}/manifest.v2.json`;
}
exports.getOperationManifestUrl = getOperationManifestUrl;
function signatureForLogging(signature) {
    if (typeof signature !== 'string') {
        return '<non-string>';
    }
    return signature.substring(0, 8);
}
exports.signatureForLogging = signatureForLogging;
